modules/core/shared/src/main/scala/data/LTree.scala

```diff
@@ -0,0 +1,60 @@
+// Copyright (c) 2018-2021 by Rob Norris
+// This software is licensed under the MIT License (MIT).
+// For more information see LICENSE or https://opensource.org/licenses/MIT
+
+package skunk.data
+
+import cats.Eq
+
+final case class LTree private (labels: List[String]) {
+
+  def isAncestorOf(other: LTree): Boolean = {
+    labels.isEmpty ||                 // Empty LTree is parent to all
+    other.labels.startsWith(labels)   // Other labels starts with this labels
+  }
+
+  def isDescendantOf(other: LTree): Boolean = other.isAncestorOf(this)
+  
+  override def toString: String = labels.mkString(LTree.Separator.toString())
+}
+
+object LTree {
+  val Empty = LTree(Nil)
+
+  def unsafe(labels: String*): LTree = LTree(labels.toList)
+
+  def fromString(s: String): Either[String, LTree] = {
+
+    if(s.isEmpty()) {
+      Right(LTree.Empty)
+    } else {
+      // We have a failure sentinal and a helper to set it.
+      var failure: String = null
+      def fail(msg: String): Unit =
+        failure = s"ltree parse error: $msg"
+      
+      val labels = s.split(Separator).toList
+
+      if(labels.length > MaxTreeLength)
+        fail(s"ltree size (${labels.size}) must be <= $MaxTreeLength")
+
+      labels.foreach(l => l match {
+        case ValidLabelRegex() => ()
+        case _ => fail(s"invalid ltree label '$l'. Only alphanumeric characters and '_' are allowed.")
+      })
+
+      if(failure != null)
+        Left(failure)
+      else
+        Right(LTree(labels))
+    }
+  }
+
+  val MaxLabelLength = 255
+  val MaxTreeLength = 65535
+  
+  private val Separator = '.'
+  private val ValidLabelRegex = s"""^[A-Za-z0-9_]{1,$MaxLabelLength}$$""".r
```

Comment:
This is actually more restrictive than the implementation, which apparently depends on the character encoding, which Skunk requires to be UTF-8. So `foo.βar.baz` is a legal `lpath`. But unless we look at the source it's unlikely we'll be able to figure out exactly what it's looking for.

Reply:
I adjusted the label regex such that it seems to pass your example but I'm sure that doesn't necessarily meet the spec.

Reply:
Yeah the spec is probably "whatever the implementation does", which we might want to look at. It would be nice to follow the same rules on the Skunk end.

Reply:
I think [this](https://github.com/postgres/postgres/blob/c30f54ad732ca5c8762bb68bbe0f51de9137dd72/src/backend/regex/regc_pg_locale.c#L332) is the implementation used to check is the character is allowable. My C is more than rusty but I'll have a closer look this weekend.

Reply:
Also, looks like my updated regex may have broken the tests on the ScalaJS side.